enhancements/ingress/lb-subnet-selection-aws.md

```diff
@@ -0,0 +1,592 @@
+    reason: OperandsProgressing
+    status: "True"
+    type: LoadBalancerProgressing
+```
+
+We have intentionally designed this feature to require manual deletion of the
+service for the following reasons:
+
+##### Reason #1: To mitigate risks associated with cluster admins providing an invalid annotation value.
+
+While the AWS Cloud Controller Manager (CCM) handles invalid annotations by
+producing errors and events, it is difficult for the Ingress Operator to
+decipher these CCM events _after_ the load balancer has been provisioned.
+Therefore, the Ingress Operator can't indicate to a cluster admin that an
+IngressController's load balancer is in a malfunctioning state (i.e., the
+service is not getting reconciled) while the service is already provisioned.
+
+When the IngressController is first created and the LoadBalancer-type
+service is not yet provisioned, these same invalid `Subnets` values will
+prevent the load balancer from being provisioned. The existing Ingress
+Operator logic will clearly indicate to the cluster admin that the load balancer
+failed to provision via the `LoadBalancerReady` status condition. In addition,
+the `LoadBalancerReady` condition will include the CCM event logs that indicate
+to the cluster admin that the `Subnets` value is invalid (see [Support Procedures](#support-procedures)
+for examples). Therefore, requiring manual service deletion provides a way
+for the Ingress Operator to produce a clear signal to the cluster admin
+that `Subnets` is invalid.
+
+See [Invalid Subnet Annotation Values](#invalid-subnet-annotation-values) for
+examples of invalid subnet annotation values.
+
+##### Reason #2: To mitigate upgrade compatibility issues.
+
+**Note**: Directly configuring the `service.beta.kubernetes.io/aws-load-balancer-subnets`
+on the LoadBalancer-type service is not supported and has never
+been. However, this enhancement is designed to prevent cluster disruption upon
+upgrading with this unsupported configuration.
+
+If the `service.beta.kubernetes.io/aws-load-balancer-subnets` annotation was
+previously configured and the cluster is upgraded, the default value `[]` for
+`Subnets` will clear the annotation for existing IngressControllers, which
+would break cluster ingress. However, requiring the service to be deleted to effectuate
+a `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets`
+value different from the current value of the annotation prevents this
+automatic removal of the service annotation on upgrade.
```

Comment:
You want to have a distinction between "no opinion" and "I have an opinion, the list should be empty"? That can be achieved in the API if you were to make the API a pointer to a slice. When the field is not present in the serialized form the value will be nil, when the empty slice is present in the serialized form, that would create you a pointer to an empty slice.

Given that, you could handle the upgrade case without requiring the manual deletion.

Reply:
I responded in [Slack](https://redhat-internal.slack.com/archives/CE4L0F143/p1714141413339429?thread_ts=1711471549.819729&cid=CE4L0F143), but responding here for completeness.

Decision for use of pointer is moot to solve upgrade problem because we still either need immutability or delete-service-to-effectuate to solve the "updating the subnets after service is created doesn't work reliably issue". I think that should be the question we answer (and sounds like we are landing on delete-service-to-effectuate).